M2/Macaulay2/packages/GKMManifolds.m2

```diff
@@ -0,0 +1,1468 @@
+-----------------------------------------------------------------------------------------------
+
+--a MomentGraph G is a MutableHashTable with the minimal data of:
+--G.vertices : a list representing vertices of the graph G
+--G.edges : a hash table whose keys are pairs {p,q} of vertices and values v are directed weights
+-- of the edge  p --> q where v is a list of numbers such that matrix{v} * basis(1,G.HTpt)
+-- gives the weight
+--G.HTpt : a polynomial ring, represent the T-equivariant cohomology ring of a point
+
+
+MomentGraph = new Type of MutableHashTable
+
+MomentGraph.synonym = "moment graph"
+
+globalAssignment MomentGraph
+net MomentGraph := G -> net ofClass class G | " on " | toString(#G.vertices) | " vertices with " | toString(#G.edges) | " edges "
+
+--given a list V representing the vertices, E a hash table of edges, and H a ring for HTpt
+--outputs a moment graph with such data
+momentGraph = method()
+momentGraph(List,HashTable,Ring) := MomentGraph => (V,E,H) -> (
+    new MomentGraph from {
+	symbol vertices => V,
+	symbol edges => E,
+	symbol HTpt => H,
+	cache => new CacheTable
+	}
+    )
+
+
+--outputs the graph underlying a moment graph G
+underlyingGraph(MomentGraph) := Graph => G -> graph(G.vertices, keys G.edges, EntryMode => "edges")
+
+
+--If a moment graph G comes from a possibly singular GKM variety with a T-invariant
+-- Whitney stratification consisting of affine spaces,
+--the vertices of G correspond to each strata,
+--and are ordered v1 <= v2 if the closure of stratum of v2 is contained in that of v1.
+cellOrder = method()
+cellOrder(MomentGraph,Poset) := (G,P) -> (
+    if G.cache.?cellOrder then (
+	print "warning: overwriting a previously defined cell order on this moment graph "
+	);
+    if G.vertices =!= P.GroundSet then (
+	<< "the ground set of the poset is not the vertices of this moment graph " <<
+	return error
+	);
+    G.cache.cellOrder = P;
+    )
+
+
+cellOrder(MomentGraph) := Poset => G -> (
+    if G.cache.?cellOrder then G.cache.cellOrder
+    else << " no cell order defined on this moment graph" << return error
+    )
+
+--------------------------------------< T-varieties >-------------------------------------------
+
+
+--for setting up a character ring: given n, outputs the character ring of T = (kk^*)^n
+makeCharRing = method();
+makeCharRing(ZZ) := Ring => n -> (
+    T := symbol T;
+    ZZ[T_0..T_(n-1), MonomialOrder => GLex, Inverses=>true]
+)
+
+
+--a TVariety X is a HashTable with minimal data of:
+--X.points : a list representing torus-invariant points
+--X.charRing : a ring representing the character ring of the torus
+--additionally, a TVariety X can have:
+--X.charts : a hash table whose keys are X.points and values are (negative of) the characters 
+--of the affine chart at the point
+--X.momentGraph : a MomentGraph which is the moment graph of the T-variety
+TVariety = new Type of MutableHashTable
```

Comment:
Does this need to be mutable?

Reply:
We decided TVariety to be MutableHashTable (sort of like Polyhedron).  Not all TVariety has keys like charts or momentGraph defined, but we wanted the users to be able to view them with "peek" and also change/update previously entered "charts" (for example).  So, TVariety has been kept as MutableHashTable.